### DISCUSSION TITLE
How to concatenate strings inside hashFiles

### DISCUSSION THREAD

**Discussion by tomgrossman** [community/community#25718] [Actions]
<p>I want to use an input variable, like from a <code>matrix</code> or <code>outputs</code> of a previous step, concatenate it with a glob pattern like <code>/**</code> and use it inside the <code>hashFiles</code> expression in the <code>actions/cache@v2</code> step.<br>
For example, here is my example simple workflow yaml:</p>
<pre><code class="lang-auto">name: Matrix test
on:
  workflow_dispatch:
  push:
    branches:
      - tom-plat-matrix
    paths:
      - '.github/workflows/matrix_test.yml'
jobs:
  build-gulp-apps:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        apps: [
          {
            name: 'static',
            path: './src/servers/static'
          },
        ]
    steps:
      - uses: actions/checkout@v2
      - uses: actions/cache@v2
        with:
          key: ${{ hashFiles(matrix.apps.path/**) }}
          path: ${{ matrix.apps.path }}/js/dist
      - if: steps.cache.outputs.cache-hit != 'true'
        run: npx gulp ${{ matrix.apps.name }}-scripts
</code></pre>
<p>However, the workflow doesn’t run due to the following error:</p>
<blockquote>
<p>The workflow is not valid. .github/workflows/matrix_test.yml (Line: 24, Col: 16): Unexpected symbol: ‘path/**’. Located at position 23 within expression: hashFiles(matrix.apps.path/**)</p>
</blockquote>
<p>That makes sense since there is no <code>path/**</code> inside the matrix’ apps object.<br>
I tried using <code>join</code> and <code>format</code> expressions, but that’s not possible inside the <code>hashFiles</code> expression.<br>
Eventually I solved it by generating this string in a previous step to the cache step:</p>
<pre><code class="lang-auto">name: output ${{ matrix.apps.name }} glob
id: glob
run: |
  echo "::set-output name=glob::${{ matrix.apps.path }}/**"
</code></pre>
<p>and using the <code>steps.glob.outputs.glob</code> in the cache key instead.<br>
However, that seems weird to me that it’s no solvable in a single cache step. This is just a simple example of a much more complex matrix, jobs and steps. If every time I need to add something to a matrix/input string I will need to create another step before, it would make my workflow file a lot longer and complicated.<br>
So, what did I miss here?</p>

**Comment by bcowdery** [comment, marked as answer]
<p>We do something similar for caching our C# nuget packages. Since we use a mono-repo we need the <code>hashFiles</code> key to be based off the project sub-directory. You can use the <code>format</code> helper to build the path strings - Try <code>hashFiles(format('{0}/**', matrix.apps.path))</code>.</p>
<p>Here’s our cache action for reference.</p>
<pre><code class="lang-auto">    - name: Cache NuGet packages
      uses: actions/cache@v2
      with:
        path: ~/.nuget/packages
        key: ${{ runner.os }}-nuget-${{ hashFiles(format('{0}/**/*.csproj', env.PROJECT_DIR), format('{0}/*.props', env.PROJECT_DIR)) }}
        restore-keys: ${{ runner.os }}-nuget
</code></pre>

**Comment by tomgrossman** [comment]
<p>wow that works! I swear I tried the <code>format</code> option as I wrote in the original question, but I guess I was dreaming? :thinking: weird that <code>join</code> doesn’t work as well, both are similar expressions. But <code>format</code> will do! Thanks!</p>
